[ 2주차 기본/심화/공유 과제 ] 웨비의 쇼핑몰

imddoy opened the pull request:
<!-- PR의 제목은 "[ n주차 기본/심화/공유 과제 ] 과제명 " 으로 작성해주시면 되겠습니다 -->
<!-- PR은 리뷰어를 위한 글입니다, 보다 더 상세하게 적음으로써 소통을 활성화해봐요! -->

## ✨ 구현 기능 명세
### 🧩 기본 과제
1. **header **
    - [x] 아이콘 클릭시 home으로 이동

2. **nav 구현 (목차)**
    - [x] 클릭 시 카테고리별 상품 리스트 보여줍니다. 
    1. 전체 → 모든 item 렌더링
    2. 카테고리 → 카테고리별 필터링
    
3. **section 구현**
    - [x] 첫 화면에선 전체 리스트가 렌더링됩니다.
    - [x] nav를 클릭할 때 마다 해당하는 리스트를 필터링한 후 렌더링 합니다.
    - [x] item 클릭시 장바구니에 담기 alert가 뜨고 yes누르면 장바구니로 이동됨
    - [x] 최초데이터 데이터 상수 파일 생성
    
4. **우측 sidebar 구현 (navigation)**
    - [x] 햄버거 아이콘 클릭 시 오른쪽 메뉴가 보이도록 구현
    - [x] 관심상품 보기, 장바구니, 관리자 3가지가 포함된 nav를 구현 (hover적용)
    - [x] 메뉴 왼쪽 상단 메뉴를 닫는 버튼 구현 
    - [x] 장바구니 페이지 클릭시 장바구니 페이지로 이동
    
5. **장바구니 페이지 구현**
    - [x] 장바구니에 담긴 cart 리스트 구현 
    - [x] 이미지, 상품명, 가격, 카테고리가 포함되게 구현
    - [x] 삭제버튼 클릭시 장바구니에서 삭제기능 구현.
    - [x] 장바구니 리스트 일정 높이가 넘어갈 시에 스크롤이 생기도록 구현 (영상 첨부!)
    - [x] 상품 금액엔 3자리마다 , 를 찍어 구분
    - [x] 홈버튼 클릭시 홈으로 이동하는 기능 구현
   
6. **구매 모달 구현**
    - [x] 구매하기 버튼 클릭시 구매를 확정짓는 모달을 뜨도록 구현
    - [x] 모달 내 구매 버튼 클릭시 alert메세지와 함께 모달 닫기
    - [x] 닫기 버튼 구현
    - [x] 장바구니에 담긴 아이템들의 총 금액 보여주기
    - [x] 상품 금액엔 3자리마다 , 를 찍어 구분 
    
   

---
### 🔥 심화 과제
1. **nav**
   - [x] nav모달이 부드럽게 나타나도록 animation 적용
    
2. ****장바구니 페이지****
   - [x] 각 장바구니 아이템에 input 구현
   - [x] 체크된 item들만 구매하기 모달에 들어가도록 구현
   - [x] 최상단에 전체 체크가 가능한 input을 만들어 전체 상품을 한번에 구매/취소할 수 있도록 구현			 
	 - [x] 체크된 항목을 구매모달을 통해 구매하면 장바구니 목록에서 사라지도록 구현

        
### 공유과제
- [ ] 명령형 프로그래밍과 선언형 프로그래밍
- [x] localStorage와 sessionStorage

***링크 첨부*** : https://imdevdoy.notion.site/local-storage-session-storage-cookie-a2f89af40bc045bfa3213e13a86bfd4d?pvs=4

<br />

##  📌 내가 새로 알게 된 부분
<!-- 새롭게 알게 된 부분 가볍게 기록하기 (기록하면서 개발하기!) -->
- dataset
체크된 요소는 가져오는데, 어떻게 체크된 요소에게서 아이템 정보를 가져올 수 있을까에 대해 계속 고민하고 공부하다가
dataset을 처음 알게 되었고 이번 기회에 다뤄보게 되었습니다.
``` js
// 장바구니 목록 렌더링
// ...
const tr = document.createElement("tr");
    // 데이터셋에 아이템 속성을 개별적으로 추가
    tr.dataset.id = item.id;
    tr.dataset.name = item.name;
    tr.dataset.image = item.image;
    tr.dataset.price = item.price;
    tr.dataset.category = item.category;
// ...

// 구매 목록에 아이템 추가
function addBuyCart(item) {
  let hasCheckedItems = false; // 체크된 아이템 여부
  checkboxes.forEach((checkbox) => {
    if (checkbox.checked) {
      hasCheckedItems = true;
      const tr = checkbox.closest("tr"); // 체크된 tr 가져오기
      const item = {
        id: tr.dataset.id,
        name: tr.dataset.name,
        image: tr.dataset.image,
        price: tr.dataset.price,
        category: tr.dataset.category,
      };
      buyList.push(item);
    }
  });
  return hasCheckedItems;
}
```

<br />

## 💎 구현과정에서의 고민과정(어려웠던 부분) 공유!
- 테이블 스크롤
테이블 스크롤을 구현하려고 tbody에 `display: block; max-height: 35rem; overflow-y: auto;`를 하니까
`display: block`때문에 표로 구현이 안되고 첫번째 칸에 내용들이 엄청 낑겼습니다.
<img width="602" alt="image" src="https://github.com/NOW-SOPT-WEB/ChaeHyeonKim/assets/90364711/946ceecb-0432-4ce7-9d87-b9bfec58d1c0">

아래 코드를 작성해서 해결을 하긴 했는데요
``` css
thead,
tbody tr {
  display: table;
  width: 100%;
}
```
<img width="935" alt="image" src="https://github.com/NOW-SOPT-WEB/ChaeHyeonKim/assets/90364711/bc1021c1-0573-402e-ad2e-8597031afd2c">

표처럼 레이아웃이 지켜져야 하는데, 벽돌처럼 살짝 어긋나져 있어서 너무 킹받아서 계속 검색해봤는데
딱히 해결이 되지 않았습니다......
원래 html 코드에서 th에 colspan을 주었는데 이것이 원인이라고 생각되었습니다.
그래서 colspan을 없애고 정말 하드코딩으로 해결했습니다.
``` css
thead th:nth-child(1),
tr td:nth-child(1) {
  width: 10%;
}
thead th:nth-child(2) {
  width: 50%;
}
tr td:nth-child(2) {
  width: 15%;
}
tr td:nth-child(3) {
  width: 35%;
}
thead th:nth-child(3),
tr td:nth-child(4) {
  width: 15%;
}
thead th:nth-child(4),
tr td:nth-child(5) {
  width: 15%;
}
thead th:nth-child(5),
tr td:nth-child(6) {
  width: 10%;
}
```
<img width="932" alt="image" src="https://github.com/NOW-SOPT-WEB/ChaeHyeonKim/assets/90364711/56e31631-270d-4d64-8bea-22d89551f3cc">

겉으로는 아무 이상이 없어 보이는데, 코드가 너무 더러워졌는데,,,,
이거 어떻게 해결해야 할까요...?

- Number 사용
id값을 비교해서 필터링을 하는 코드를 작성하였는데, 계속 오류가 나서 콘솔창에 type을 출력해보니 id를 문자열로 가져와서 비교가 제대로 되지 않았습니다.
로컬 스토리지가 문자열로 저장이 되었기 때문이라고 생각했고 Number를 사용해서 제대로 비교가 될 수 있도록 해주었습니다.
``` js
// 구매 확정
function goConfirm() {
  let cartList = JSON.parse(localStorage.getItem("cartList") || "[]");
  const buyIds = buyList.map((buy) => Number(buy.id)); // 구매 목록의 id 배열 생성
  cartList = cartList.filter((cart) => !buyIds.includes(cart.id)); // 구매 목록에 없는 상품들만 남김
  localStorage.setItem("cartList", JSON.stringify(cartList)); // 변경된 장바구니 목록을 로컬 스토리지에 저장
  alert("구매가 성공적으로 완료되었습니다.");
  location.href = location.href; // 페이지 새로고침
}
```
이 코드 이외에도 계속해서 Number를 사용해서 오류를 해결했는데, 솔직히 많이 번거롭기도 하고 올바른 해결이라고 생각이 안되었습니다. 이렇게 해결을 해도 되는 걸까요...?

- js를 어떻게 나눠야 할지 모르겠어서 한 파일로 그냥 길게 쭈욱 썼는데 어떻게 나눠야 할지 추천 부탁드립니다 🙏🏻🙏🏻

<br />

## 🥺 소요 시간
- `12h`

<br />

## 🌈 구현 결과물
https://sopt-week2-shoppinglist.vercel.app/

imddoy committed:
init: 2차 세미나 과제 초기세팅

imddoy committed:
refactor: header 분리

imddoy committed:
feat: 헤더 아이콘 클릭시 home 이동

imddoy committed:
refactor: 헤더 구현 변경
fetch api는  지연이 발생할 수 있으므로 웹 컴포넌트로 방법 변경

imddoy committed:
feat: index 페이지 상품 데이터 렌더링 및 필터링

imddoy committed:
feat: 장바구니 추가

imddoy committed:
feat: 장바구니 추가 및 삭제
로컬스토리지 이용해서 추가 및 개별 삭제 구현

imddoy committed:
feat: input 박스 구현 및 모달창 퍼블리싱
input 전체 선택 및 해제 구현
모달창 오픈 및 클로즈 구현

imddoy committed:
feat: 구매하기 구현 완료
장바구니 목록에서 구매 목록 삭제
구매 완료시 페이지 새로 렌더링(체크 박스, 장바구니 목록 등)

imddoy committed:
feat: 총 구매 금액 출력

imddoy committed:
feat: 테이블 스크롤

imddoy committed:
feat: 사이드바 구현

imddoy committed:
design: 푸터 고정

imddoy committed:
cleanup: 중복 코드 삭제

pepperdad commented on week2_shoppinglist/index.html:
```diff
@@ -0,0 +1,61 @@
+  <head>
+    <meta charset="UTF-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <meta property="og:title" content="잔망루피 쇼핑리스트" />
+    <meta
+      property="og:description"
+      content="새싹웹팟 2차 세미나 과제 - 쇼핑리스트 만들기"
+    />
+    <link rel="icon" type="image/png" href="images/favicon.jpg" />
+    <meta
+      property="og:image"
+      content="https://sopt-week1-shoppinglist.vercel.app/images/logo.png"
+    />
```
p5: 메타데이터 👍👍👍

suwonthugger replied:
p5: 와우 메타데이터 정말 최고!

HADONGMI replied:
메타데이터 첨봤는데 데이터에 대한 데이터로, 다른 데이터를 설명하는 정보이고,웹 개발의 맥락에서, HTML 문서 내의 메타 데이터는 웹 페이지의 내용을 설명하는 정보를 제공하더라구여

새로운거 하나 배워갑니다~~

pepperdad commented on week2_shoppinglist/index.js:
```diff
@@ -0,0 +1,384 @@
+  cartList.push(item); // 새 상품 추가
+  localStorage.setItem("cartList", JSON.stringify(cartList)); // 변경된 장바구니 목록을 로컬 스토리지에 저장
+  alert(`${item.name}을(를) 장바구니에 담았습니다.`);
+  var answer = confirm("장바구니로 이동하시겠습니까?");
```
p3: var는 웬만하면 안 쓰는 게 좋아! 여기서는 const로 바꾸는 건 어떨까!?
```suggestion
  const answer = confirm("장바구니로 이동하시겠습니까?");
```

HADONGMI replied:
var, let, const 차이점도 같이 살펴보면 좋을 거같아염ㅎㅎ

https://velog.io/@bathingape/JavaScript-var-let-const-%EC%B0%A8%EC%9D%B4%EC%A0%90

imddoy replied:
오오...! 벨로그까지!! 제안 감사합니다! 수정했습니다!

pepperdad commented on week2_shoppinglist/index.js:
```diff
@@ -0,0 +1,384 @@
+/* header */
+
+// 웹 컴포넌트로 header 삽입
+class HeaderComponent extends HTMLElement {
+  connectedCallback() {
+    this.innerHTML = `
+    <header>
+      <img
+        src="./images/logo.png"
+        alt="쇼핑몰 로고"
+        width="36rem"
+        onclick="goHome()"
+      />
+      <h1 lang="en">SHOPPING LIST</h1>
+      <img src="./images/menu.svg" alt="햄버거" class="menuBtn" onclick="openNav()" />
+    </header>
+    <nav class="sidebar">
+      <div class="menuBtn" onclick="closeNav()">x
+      </div>
+      <ul>
+        <li>관심 상품 목록</li>
+        <li>관리자 페이지</li>
+        <li><a href="cart.html">장바구니</a></li>
+      </ul>
+    </nav>
+      `;
+  }
+}
+customElements.define("shop-header", HeaderComponent); // html 태그 정의 및 구현
```
p5: 우와 이렇게 컴포넌트를 만들 수 있구나... 
좋은 정보 알아갑니다 ㅎㅎㅎ

suwonthugger replied:
p5: 오 이런 참신한 방법도 있군요.. 많이 배워갑니다!

HADONGMI replied:
저도 새로운 하나 배워갑니당~~

imddoy replied:
저도 이번에 한번 사용해보고 싶어서 처음 시도해봤습니다...ㅎㅎ

pepperdad commented on week2_shoppinglist/index.js:
```diff
@@ -0,0 +1,384 @@
+    // 이미지
+    const img = document.createElement("img");
+    img.src = item.image;
+    img.alt = "상품 이미지";
```
p4: 여기 상품 리스트 전체에 alt 속성이 "상품 이미지"라는 string으로 들어가고있어!!! 
이 부분은 검색엔진이나 스크린 리더에서도 중요하게 사용되는 부분이여서 아래처럼 바꾸는 건 어떨까? ㅎㅎ
```suggestion
    img.alt = item.name;
```
참고: https://rgy0409.tistory.com/2884

imddoy replied:
아...!!!! 이런 방법이 있었네요!! 감사합니다🙇🏻‍♀️🙇🏻‍♀️

pepperdad commented on week2_shoppinglist/index.js:
```diff
@@ -0,0 +1,384 @@
+    const imgtd = document.createElement("td");
+    const img = document.createElement("img");
+    img.src = item.image;
+    img.alt = "상품 이미지";
```
p5: 여기도 마찬가지!

pepperdad commented on week2_shoppinglist/index.js:
```diff
@@ -0,0 +1,384 @@
+//장바구니 목록 추가
+function addCart(item) {
+  let cartList = JSON.parse(localStorage.getItem("cartList")) || []; // cartList가 배열이 아닌 경우, 빈 배열로 초기화
+  cartList.push(item); // 새 상품 추가
+  localStorage.setItem("cartList", JSON.stringify(cartList)); // 변경된 장바구니 목록을 로컬 스토리지에 저장
```
p3: 여기서 push 함수보다는 이번 공유과제에서 배웠던 배열의 불변성의 법칙을 지키면서 추가하는 건 어떨까~?

```suggestion
  const newCartList = [...cartList, item];  // 새 상품 추가
  localStorage.setItem("cartList", JSON.stringify(newCartList)); // 변경된 장바구니 목록을 로컬 스토리지에 저장
```
이렇게 할 수 있을 것 같애!

imddoy replied:
오🫢직접 코드까지 수정해주시다니!! 감사합니다!! 배열의 불변성을 항상 안지켜왔던 것 같아서 반성합니다...
이번 공유 과제를 한번 적용해보도록 하겠습니다!

buyList까지 적용했습니다!ㅎㅎ

pepperdad reviewed (commented):
js로 컴포넌트를 만들 수 있다는 유용한 사실을 알게되어... 너무 좋다...
너무너무 고생했어 🥧🥧🥧
comment 한 번 확인하고, 반영하면 좋을 것 같아!

suwonthugger commented on week2_shoppinglist/docs/README.md:
```diff
@@ -0,0 +1,47 @@
+## 나만의 쇼핑리스트 만들기
+
+### 🧩 기본 과제
+
+1. **header**
+
+   - [x] 아이콘 클릭시 home으로 이동
+
+2. **nav 구현 (목차)**
+   - [x] 클릭 시 카테고리별 상품 리스트 보여줍니다.
+   1. 전체 → 모든 item 렌더링
+   2. 카테고리 → 카테고리별 필터링
```
p5: 리드미 명세서 작성 👍

HADONGMI replied:
굿굿

suwonthugger commented on week2_shoppinglist/index.js:
```diff
@@ -0,0 +1,384 @@
+/* index */
+
+// 상품 리스트
+const SHOPPING_LIST = [
+  {
+    id: 1,
+    category: "stationery",
+    image: "./images/note.jpg",
+    name: "잔망루피 페이스 스프링 노트",
+    price: 4000,
+    liked: false,
+  },
```
p5) 상수를 분리해서 import하는건 어떠신가요~?

HADONGMI replied:
https://velog.io/@zzangzzong/React-%EC%83%81%EC%88%98%EB%8D%B0%EC%9D%B4%ED%84%B0

이거 한번 참고해보시면 좋을 거같아염ㅎㅎ

imddoy replied:
오오 다들 친절하게 가르쳐주셔서 감사합니다ㅎㅎ
확실히 코드가 더 보기 편해졌어요~~
![SoExcited~GIF](https://github.com/NOW-SOPT-WEB/ChaeHyeonKim/assets/90364711/d90e2791-0c7c-4602-a003-b5bd64a22e7f)

suwonthugger commented on week2_shoppinglist/index.js:
```diff
@@ -0,0 +1,384 @@
+    const checkbox = document.createElement("input");
+    checkbox.type = "checkbox";
+    checkbox.name = "item";
+    checkbox.onclick = function () {
+      onclick = checkSelectAll(this);
+    };
```
p5) this 키워드는 화살표 함수, 일반 함수에 따라서 바인딩이 달라질 수 있어서 위험할 수도 있는데요~ 여기서는 event.target을 사용하시는건 어떠세요~~? 

```suggestion
checkbox.onclick = function (event) {
      onclick = checkSelectAll(event.target);
    };
```

suwonthugger commented on week2_shoppinglist/index.js:
```diff
@@ -0,0 +1,384 @@
+  cartList = cartList.filter((cart) => !buyIds.includes(cart.id)); // 구매 목록에 없는 상품들만 남김
+  localStorage.setItem("cartList", JSON.stringify(cartList)); // 변경된 장바구니 목록을 로컬 스토리지에 저장
+  alert("구매가 성공적으로 완료되었습니다.");
+  location.href = location.href; // 페이지 새로고침
```
p5: 오 새로고침을 통해서 변경된 장바구니를 다시 보여주는 방법도 있군요~

imddoy replied:
장바구니 페이지에는 장바구니 테이블밖에 없어서 새로고침으로 다시 보여줘도 문제가 없을 것 같아서 새로고침으로 구현했습니다..!

suwonthugger commented on week2_shoppinglist/images/favicon.jpg:
p5: 루피가 너무 귀엽네요~

imddoy replied:
루피가 고맙다고 전해달라네요~~

10tacion reviewed (approved):
컴포넌트를 만들어서 렌더링하는 개념을 적용하신게 너무 참신했어요~ 저도 다음에 이렇게 만들어보고 싶네요! 멋진 과제입니다!
너무 고생하셨어요 🔥

HADONGMI commented on week2_shoppinglist/cart.css:
```diff
@@ -0,0 +1,129 @@
+  display: flex;
+  gap: 1rem;
+  width: 100%;
+  padding: 2rem;
+  box-sizing: border-box;
+  overflow: auto;
+}
+.buylist::-webkit-scrollbar {
+  height: 0.3rem;
```
px보다 rem사용 좋은 것 같아여~~

https://yozm.wishket.com/magazine/detail/1410/

HADONGMI commented on week2_shoppinglist/style.css:
```diff
@@ -0,0 +1,317 @@
+  padding: 0;
+  list-style-type: none;
+}
+.sidebar li {
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  height: 3rem;
+  padding: 1rem;
+  font-size: 1.5rem;
+  color: #fff;
+  cursor: pointer;
+}
+.sidebar li:hover {
```
px 대신에 rem사용 굿이에염ㅎㅎ

imddoy committed:
fix: 변수 선언 변경
재할당이 필요하지 않는 변수는 var가 아닌 const로 선언

imddoy committed:
fix: img alt 변경
검색엔진최적화를 위해 상품명으로 변경

imddoy committed:
fix: img alt 변경
검색엔진최적화를 위해 상품명으로 변경

imddoy committed:
fix: 배열의 불변성 지키도록 배열 복제
newCartList, newBuyList에 ...연산자 사용하여 복제하여 원본 배열 직접 변경하지 않도록 수정

imddoy committed:
refactor: 상수 데이터 분리
script를 module로 타입 정하면 html에 있는 onclick 이벤트가 작동하지 않음
-> 이벤트 리스너 연결로 해결
dom에 존재하지 않으면 오류 -> && 연산자 사용으로 해결
